### PR TITLE
Bad client changes Part I

### PR DESCRIPTION
**Pull request by yashykt**
This is the first part of the changes for the bad_client framework. This change enables the large_metadata test with the bad_client framework again.

This solves the issue mentioned in #11745 by passing an array of client side validators to the bad client framework as opposed to a single validator earlier. Each iteration in the test allows sending a payload from the client and running a validator on the expected response.

This avoids the race condition where the settings frame from the server is not acknowledged by the client resulting in the server accepting metadata beyond the limits set.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPong_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__32768_2.opt.new: 1


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by markdroth** [comment]
I am unfortunately way too swamped to be able to get to this anytime soon.  Reassigning to Noah for review.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: test/core/bad_client/bad_client.h [hunk 1]
```diff
@@ -28,30 +28,65 @@
...
+ * Also execute \a server_validator in a separate thread to assert that the
+ * bytes are handled as expected.
+ */
 void grpc_run_bad_client_test(
     grpc_bad_client_server_side_validator server_validator,
-    grpc_bad_client_client_stream_validator client_validator,
-    const char* client_payload, size_t client_payload_length, uint32_t flags);
+    grpc_bad_client_arg args[], int num_args, uint32_t flags);
+
+/* A hack to let old tests work as before. In these tests, instead of an array,
```
Comment (ncteisen) [inline test/core/bad_client/bad_client.h]: Slick hack, but might be worth just changing all of the existing bad clients to use your new validation syntax for consistency sake. There are not too many bad clients. I don't feel strongly about this though, so up to you.

File: test/core/bad_client/bad_client.cc [hunk 2]
```diff
@@ -62,136 +69,172 @@ static void server_setup_transport(void* ts, grpc_transport* transport) {
...
+    grpc_endpoint_destroy(*client_fd);
+    grpc_core::ExecCtx::Get()->Flush();
+    *client_fd = nullptr;
+  }
+}
 
-  if (client_payload_length < 4 * 1024) {
-    hex = gpr_dump(client_payload, client_payload_length,
+/* Runs client side validator */
+void grpc_run_client_side_validator(grpc_bad_client_arg* arg, uint32_t flags,
```
Comment (ncteisen) [inline test/core/bad_client/bad_client.cc]: I realize that this was an existing problem, but it looks like flags is only used to signal shutdown. If that is the case it should be renamed to shutdown. If there are other future uses then it should be used like if (flags & FLAGS_SHUTDOWN) {...}

File: test/core/bad_client/bad_client.h (hunk 1 above)
Comment (yashykt) [reply test/core/bad_client/bad_client.h]: The original syntax is kind of short and nice. Keeps it simple if the test is not doing something complicated, so I actually kind of like it.

File: test/core/bad_client/bad_client.cc (hunk 2 above)
Comment (yashykt) [reply test/core/bad_client/bad_client.cc]: Currently, there are two uses for flags.
One, to indicate that the client should disconnect before the validator runs.
Second, to indicate that the client is going to write too much data which might not finish writing even after the validator returns. This ends up requiring a forceful shutdown after validator runs, so  we can avoid a SEGFAULT.

I've updated the comments and the if condition to reflect the change.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yashykt** [comment]
Known Issue: #12748

**Pull request merged**